### ISSUE TITLE
Modularization

### ISSUE THREAD

**Issue by josdejong** [josdejong/mathjs#280]
math.js is growing. This is great. It supports quite some different data types now (number, Complex, Unit, Matrix, BigNumber, ...) and a lot of functions. We have plans for more data types and more advanced functions (for example numerical functions). It will be very powerful to modularize the libary. That way you don't have to distribute the whole library with your applications, but just the portion that you actually use. 

Right now, each function of mathjs checks the input type, does type conversions if needed, and executes the corresponding logic, and throws an error in case of unsupported input types. See for example [master/.../add.js](https://github.com/josdejong/mathjs/blob/master/lib/function/arithmetic/add.js#L49). We are currently refactoring the library to use a separate library, [`typed-function`](https://github.com/josdejong/typed-function), to do the type checking, so mathjs can focus on the actual logic. This results in code like [v2/.../add.js](https://github.com/josdejong/mathjs/blob/v2/lib/function/arithmetic/add.js#L39). Now, the code is split in a separate function per type of input argument. After this refactoring, it will be relatively easy step to modularize the library. We can do this in different ways:

1.  **Separate packages**
    
    Offer different packages to the user, like `mathjs-core`, `mathjs-matrix`, `mathjs-bignumber`, `mathjs-expression`, `mathjs-numerics` etc. The default package `mathjs` can contain all available packages by default. You should be able to use multiple "sub" packages in conjunction, like using `mathjs-bignumber` and `mathjs-expression` together. Difficulty may be *which* packages we want to offer exactly, how to split the functionality. There are two options here:

    1. Keep the source code in a single project, and generate multiple packages from that.
    2. Split the source code in separate projects, each project publishing a single package.
    
2.  **Custom builds** 

    Offer just a single package like it is now, and offer a (user-friendly!) build tool to generate your own custom version of math.js. This should be easy to so both in an automated way (to integrate with your build chain) as well as by hand (using a web interfarce or command line tool). You should be able to both select coarse grained groups of functionality (like include bignumber support or not), and fine grained control of including/excluding individual functions.

    When we go for this option, we get solution 1.i almost for free, but way more customizable.

Both have pros and cons. Having just a single package may make it easy for a new user, its a simple message. In first instance you're not confronted with a bunch of packages of which you may have *no clue* which ones to load. On the other hand, it may not be a big problem if in all examples and docs you just use the main, all containing `mathjs` package.

As for the customization: just selecting from a list with sub-packages may be easier than having to get into place your own custom build. But on the other hand, a custom build gives you much more fine-grained control over what you exactly want to include, and if you're looking into this sort of optimization you will probably have a build chain anyway. And besides that, the build step can be really easy if we get it right.

What are your thoughts on this matter?

**Comment by Nekomajin42** [comment]
How about both? If you offer custom packages via build tool, you can create not only a single package but some predefined packages too, can't you?

**Comment by josdejong** [comment]
That's an interesting idea. For the "separate packages" solution I was initially thinking about splitting up the *source code* in separate projects, each having one corresponding package on npm (each one complementary to the others, no overlap.

Splitting in separate projects may not be needed at all: we could keep a single code base, and from this code base generate and publish a set of npm packages (like lodash does). Can be a number of "standard recipes", does not per se need to be a set of complementary packages. 

Sounds appealing to me.

**Comment by BigFav** [comment]
My first thought when reading this is that we would probably have to do a little of the 1st option if we decided to do the 2nd option (as a tutorial of some sort), and the 2nd option removes the problem the 1st option has (in regards to getting picky with packages to make).

**Comment by josdejong** [comment]
Yes indeed, if we keep the source code in one project and add support for custom builds (option 2), we get option 1 almost for free. We should go for this solution.

I've updated the summary above.

**Comment by josdejong** [comment]
I've been thinking about what structure will be best for modularization. First, we should have a rough idea on what use cases to support:

1. An application which just uses a couple of the functions of mathjs, like for matrices, but is not interested in the expression parser or all other functions. This application wants to exclude the largest part of the *functions*, and the expression parser and related code.
2. A calculator like app, for which the expression parser is important, but which may not be interested in supporting all data types. This application may want to exclude certain *data types*, like exclude BigNumber support or Matrix support.

For node.js applications it's enough to just not load unused parts of math.js. For browser applications it's important to be able to remove unused parts of math.js from the bundled application. There are basically two solutions for modularization:

- (a) The code itself is modular and you can load individual functions.
- (b) The code itself is not modular, but there is a build script which allows you to create a custom bundle, excluding the parts that you are not interested in. 

Now (a) is always the nicest, as it does not a special build step, it's much more straight forward. What I want to be able to do is load a custom part of math.js without needing a build step for this. Something like:

```js
// create an empty instance of mathjs (without data types and functions)
var math = require('mathjs/loader').create();

// load some data types
math.import(require('mathjs/lib/type/Complex'));        
math.import(require('mathjs/lib/type/Unit'));        
math.import(require('mathjs/lib/type/Matrix'));        

// load some functions
math.import(require('mathjs/lib/function/arithmetic/add'));
math.import(require('mathjs/lib/function/arithmetic/subtract'));
math.import(require('mathjs/lib/function/arithmetic/multiply'));
math.import(require('mathjs/lib/function/arithmetic/divide'));

// use mathjs
console.log('multiply', math.multiply(2, 3)); // output: 6
```

Many functions and some of the data types depend on configuration of the mathjs instance at hand. Therefore, most of them cannot be loaded in a static way but need to be created via a factory function, where you pass for example the config. Furthermore, if we want to be able to load individual functions, we *must* use explicit dependencies everywhere. That means that we can't pass a global `math` namespace to all functions, where they can add themselves and consume whatever dependencies they need. Instead we have to `require` all dependencies explicitly.

The solution that I have in mint now looks as follows. The structure of a function looks like:

```js
var collection = require('../../type/collection');

// a factory function for constructing a cube function
function factory (type, config, load, typed) {
  // explicit dependencies, instantiate via `load` (which caches created instances)
  var multiply = load(require('./multiply'));
  
  // generate (compose) a typed-function from a number of signatures
  var cube = typed('cube', {
    'number': function (x) {
      return x * x * x;
    },

    'Complex': function (x) {
      return multiply(multiply(x, x), x);
    },
    
    'BigNumber': function (x) {
      return x.times(x).times(x);
    },
    
    'Array | Matrix': function (x) {
      return collection.deepMap(x, cube);
    }
  });
  
  return cube;
}

// export properties such that math.import can recognize this as a factory function
exports.type = 'function';
exports.name = 'cube';
exports.factory = factory;
```

Here:

- The script exports a factory function, which is recognized by exporting three properties: a `type`, a `name`, and a `factory`. The `math.import` function recognizes these and loads them correctly by passing these arguments to their factory funtion.

- The factory function is called with four arguments:
  - `type` contains data types like `Matrix`, `BigNumber`, etc. 
  - `config` contains the configuration of the current mathjs instance. 
  - `load` creates a function instance from a factory function returned by `require`, and caches them once created (i.e. when `multiply` is used by multiple functions, it will be loaded only once).
  - `typed` is a function for generating a [typed function](https://github.com/josdejong/typed-function) from a set of signatures. The typed-function will do all the type checking and type conversion stuff. This argument could be loaded like a regular function too, but it's quite an integral part of math.js v2, and saves a line of code in almost every function by just passing it as argument. 

  I have been thinking whether it's more neat to pass these arguments as a single object, but then you get much longer references to actually use them, resulting in more bloat in the code.

- Because of the explicit dependencies like `var multiply = load(require('./multiply'));`, you can actually load a subset of the functions without having a chance of missing dependencies.

- Explicit dependencies for data types will not work, as most functions contains support for all data types and hence would load all data types. So instead, data types should be loaded first and passed to the functions on instantiation (via the `type` argument). When certain data types are missing, the `typed` function should neatly ignore function signatures with this data type from the generated typed-function.

- In order to support generating a bundle with a subset of functionality (solution (b)), we can create a loader for webpack and a plugin for browserify which can dynamically generate an custom index file with just the functions and data types that you need, and also strips the unused parts from functions. With a regexp you could remove signatures from typed functions like `'BigNumber': function (x) {...}` when BigNumber is excluded. When bundling and minifying, any left over dead code will be removed. So the final bundle will really only contain the code that you actually use.

I've refactored the functions in the v2 branch to this new structure. It works quite nicely. Though the factory function structure is relatively complicated at first sight, the resulting code looks much cleaner than in v1 (compare for example [exp v1](https://github.com/josdejong/mathjs/blob/master/lib/function/arithmetic/exp.js) and [exp v2](https://github.com/josdejong/mathjs/blob/v2/lib/function/arithmetic/exp.js)).

What do you think?

**Comment by rjbaucells** [comment]
Hi Jos,

I started looking at the v2 branch implementation in order to merge the Sparse Matrix changes into this branch. I think the code is much cleaner using the `typed` functionality. 

You have not migrated any of the types (Matrix, Complex, ...) into the new file structure. You mentioned that those types will be available in the first parameter of the factory function, but later you said "When certain data types are missing...". In what scenario is that possible? I just cannot imagine one now... It would be good seeing one of the types migrated to the new file structure for the work I am planning to do with Sparse Matrices.

Did you evaluate using RequireJS as the import vehicle for the library? I have had good experiences with RequireJS for that purpose.

I have a comment related to executing the tests in this branch. I do not know it it is my imagination or not but the tests seems to run slower than in the v1 branch.

**Comment by josdejong** [comment]
Thanks for your feedback @rjbaucells, good to hear you find this new structure looking cleaner too. Indeed, I haven't migrated the datatypes (nor constants) to this new structure.

What I meant with data types missing from the `type` argument passed to the factory functions is that I want to have this `type` object (created and passed by the loader) be empty by default, and populate it using `math.import(require('mathjs/lib/type/Matrix'));`, the same way you can now load functions. This is not yet the case though, right now the loader loads all datatypes immediately.
When some of the data types are *not* loaded (say `BigNumber` is not loaded), the `typed-function` constructor should ignore signatures with these datatypes, so you get proper errors like "Unknown datatype" instead of vague errors down the wire due to `type.BigNumber` being undefined.

Yes I have experience with AMD/RequireJS. It's a powerful solution, but I personally don't like it that much, I like Commonjs more. Besides syntactical differences, do you see any advantages of AMD over Commonjs here?

The tests run indeed noticeably slower. Right now v2 loads almost twice as slow as v1 (like 260ms instead of 160ms). Also memory usage is almost twice as large. Execution of methods is slower too. This is all because we are switching from "hand optimized" methods in v1 to "automatically composed" functions in v2 (composed at runtime using typed-function). 
The current state is not yet acceptable in my opinion, I have to improve `typed-function` drastically. I expect that execution speed, which is most important here, can/should become similar to that of v1 for individual methods. Load time and memory usage will always be a bit larger than of v1, but I would be very happy if we can limit the increased load time and memory usage to say 25%, given everything we gain with it (maintainable modularized code, fast matrix calculations, etc). `typed-function` is brand new and totally non-optimized right now, so I'm optimistic here that we should be able to improve it quite a bit.

**Comment by rjbaucells** [comment]
Jos, I looked at the code in more details and I get your point on loading types. It seems to me that the `math.import()` function is not prepared to load types yet; it is loading only factories with `type === 'function'` and into the `math` namespace directly (I do not see support for loading types into `math.type` namespace).

I started cleaning up the way to port the Sparse Matrices into v2, so far I have only replaced the `new Matrix()` statements with `math.matrix()` calls. I will need support for loading types soon to port the rest of the implementation. I will wait until you have it ready, it would be great if you can work on this feature so I can work in parallel.

**Comment by josdejong** [comment]
Indeed, `math.import` still misses support for importing types. I will extend that asap, so you can go ahead with porting sparse matrices to v2. I expect to have time for this Thursday night.

**Comment by josdejong** [comment]
@rjbaucells data types can now also be refactored into factory functions. I've already refactored BigNumber this way, the other data types still need to be converted. `loader.js` is already way cleaner but still a lot of work to do.

**Comment by rjbaucells** [comment]
Thanks; expect a big push request, I am in the process of migrating type/collection that will migrate almost all files to v2. After that I will submit the sparse matrix implementation.

**Comment by josdejong** [comment]
Ok sounds great. So I suppose you already did merge v1.5 into v2? (I expect that gives some merging conflicts, haven't done it myself yet).

**Comment by rjbaucells** [comment]
I did try merging 1.5 into 2, there were too many conflicts so I decided to manually migrate `collection` to the new file structure and using `typed()`. This migration touches lot of files, after that is completed I will manually bring the sparse matrices into v2.

**Comment by josdejong** [comment]
Ok, I will do the merge from v1.5 to v2 then, must be done anyway. Does it make sense for you to await this? I expect to be able to do the merge coming Monday or Tuesday evening.

**Comment by rjbaucells** [comment]
Ok, I will wait for the merge. I tried once again and there were too many conflicts, I went to manually merge them but it seems I committed some mistakes there and the resulting code does not work at all. Do the merge, I can manually fix the conflicts related to the sparse matrix implementation. I added enough tests there so in case of a problem we will catch it very easy.

**Comment by nateabele** [comment]
> Yes I have experience with AMD/RequireJS. It's a powerful solution, but I personally don't like it that much, I like Commonjs more. Besides syntactical differences, do you see any advantages of AMD over Commonjs here?

Perhaps this idea is a bit more 'extreme', but what about writing the main source as ES6 modules? Then you could distribute AMD or CommonJS or both.

**Comment by josdejong** [comment]
ES6 modules can be a next step, I'm definitely for moving to ES6 modules, but let's not do too much at once. CommonJS syntax is closest to that of ES6 modules, so to convert from CommonJS modules to ES6 modules is quite straight forward.

**Comment by josdejong** [comment]
@rjbaucells I'm almost done with merging v1 into v2, will finish it tomorrow night. It's quite a job...

**Comment by nateabele** [comment]
> [S]o to convert from CommonJS modules to ES6 modules is quite straight forward.

:+1:

**Comment by josdejong** [comment]
@rjbaucells I've merged v1.5 into v2. The `format` support of the functions `eye` and `diag` still need to be fixed, can you do that for me? (is probably faster to do for you than for me). You will find `FIXME`'s in the code and unit tests. 

You will see the signatures of these two functions with `typed-function` are quite verbose. We need to implement optional arguments in `typed-function` to simplify it.

**Comment by rjbaucells** [comment]
Thanks, I will review it and make the appropriate changes.

**Comment by rjbaucells** [comment]
@josdejong I completed the fixes for `eye` and `diag` in v2. I also found some of the functions that did not have the improvements for sparse matrices. I will continue re-viewing the v2 branch in case I find any other place that need changes.

**Comment by josdejong** [comment]
Thanks a lot!

**Comment by josdejong** [comment]
@rjbaucells and I had a good discussion in #321 about how to group and strore functionality. Basically, there are two options:

1. Group functionality per *type*, like writing all functions for Matrix and BigNumber as member functions of the prototype.
  - Pros: It's very natural to store the functions acting on a data type with the data type itself in a prototype. It's a logical grouping.
  - Cons: Results in a limited number of huge files, which isn't good for modularity of the code.

2. Group functionality per *function*. For example the function `add` will contain all implementations of the function for different data types (number, BigNumber, Matrix, sparse matrix, etc). 
  - Pros: assuming that we have "tens" of data types and "hundreds" of functions, it allows for more modular code when grouping per function than per data type. This way you can easily create custom bundles containing a limited number of functions/data types.  
  - Cons: code related to a single data type is scattered around lots of files. Extra "util" files needed to store functionality shared by multiple functions. The functions need to access internal properties of the data types, which is not so neat.

For me the modularity argument is *very* important, and this is what we aim for with `v2`. I hope math.js will keep growing and growing. That means though that in the end the average user only uses a tiny fraction of the library. It's important to make it a no-brainer to load and/or bundle only the used part, instead of having to distribute costly kB's over the wire, or have to wait for seconds before the library is loaded and having it consume lots of extra memory.

Therefore I propose to go for grouping functionality per *function* rather than per *data type*. A more *functional* than *object oriented* approach. Thus, as much as possible static, pure functions, and as little as possible member functions. We can go this way and after half a year or so evaluate whether this results in a practical solution with the desired modularity.

What are your thoughts on this matter?

**Comment by nateabele** [comment]
There's also the Haskell typeclass approach of treating a type definition similar to an interface that must be implemented: the type defines how it handles low-level atomic operations (addition, subtraction, comparison, etc.), and specifies an identity value.

If a common set of these low-level type operations needed across all functions could be identified, it should be possible for types and functions to be fully decoupled, and fully composable.

**Comment by josdejong** [comment]
@nateabele thanks for your suggestion. I don't yet fully get your point. Can you elaborate a little bit more on how this would look like in JavaScript (which does not have interfaces)?

**Comment by josdejong** [comment]
Ah, I just read the [introduction of this chapter on typeclasses in Haskell](http://book.realworldhaskell.org/read/using-typeclasses.html), I think I get the point. 

This is one of the things I want to achieve with `typed-function`, which allows to merge multiple typed-functions into one, like:

```js
var add1 = typed('add', {
  'number, number': function (a, b) {
    return a + b;
  });

var add2 = typed('add', {
  'Complex, Complex': function (a, b) {
    return new Complex(a.re + b.re, a.im + b.im);
  });

// the merged function `add` can handle both numbers and Complex
var add = typed(add1, add2); 
```
Now this `merge` action can be easily/automatically be done by the `math.import` function.

I'm only a little bit in doubt about the explosion in numbers of files and the performance implications of this. Anyway `v2` is already heading in this direction.

**Comment by nateabele** [comment]
:+1:

**Comment by josdejong** [comment]
I've now refactored all functions (except for some details). Hurray. It works out quite nicely so far! 

There are still some improvements needed in `typed-function` before we can fully embrace the new modular structure.

I will now reorganize the files a little bit so it will be easy to load related functionality with a single import statement. Roughly:
- group all matrix related types and function in one folder `./lib/type/matrix/*`, move collection.js there too.
- move all data types into a sub folder of their own and move the corresponding construction functions there. 
- move `./lib/function/expression/*` to `./lib/expression/function/*`  
- put all files related to the core in a single folder `./core` (import.js, config.js, core.js).

So if you're missing files they probably have been moved somewhere else.
